dnsext-svcb/DNS/SVCB.hs

```diff
@@ -0,0 +1,114 @@
+{-# LANGUAGE PatternSynonyms #-}
+{-# LANGUAGE RecordWildCards #-}
+
+-- | This module provides Service Binding (SVCB) RR and HTTPS RR.
+module DNS.SVCB (
+  -- * Extension
+    addResourceDataForSVCB
+  -- ** DNS types
+  , TYPE (
+    SVCB
+  , HTTPS
+  )
+  -- ** Resource data
+  , RD_SVCB(..)
+  , RD_HTTPS(..)
+  -- * Service parameters
+  , SvcParams
+  , lookupSvcParams
+  -- ** Service parameter keys
+  , SvcParamKey (
+    SPK_Mandatory
+  , SPK_ALPN
+  , SPK_NoDefaultALPN
+  , SPK_Port
+  , SPK_IPv4Hint
+  , SPK_ECH
+  , SPK_IPv6Hint
+  )
+  , fromSvcParamKey
+  , toSvcParamKey
+  -- ** Service parameter values
+  , SvcParamValue
+  , SPV(..)
+  , SPV_Mandatory(..)
+  , SPV_Port(..)
+  , SPV_IPv4Hint(..)
+  , SPV_IPv6Hint(..)
+  , SPV_Opaque(..)
+  ) where
+
+import DNS.Types
+import DNS.Types.Internal
+import qualified DNS.Types.Opaque as Opaque
+import qualified Data.IntMap as M
+
+import DNS.SVCB.Imports
+import DNS.SVCB.Key
+import DNS.SVCB.Params
+import DNS.SVCB.Value
+
+----------------------------------------------------------------
+
+pattern SVCB :: TYPE
+pattern SVCB  = TYPE 64
+
+pattern HTTPS :: TYPE
+pattern HTTPS = TYPE 65
+
+----------------------------------------------------------------
+
+data RD_SVCB = RD_SVCB {
+    svcb_priority :: Word16
+  , svcb_target   :: Domain
+  , svcb_params   :: SvcParams
+  } deriving (Eq,Ord,Show)
+
+instance ResourceData RD_SVCB where
+    resourceDataType _ = SVCB
+    putResourceData _ RD_SVCB{..} = do
+        put16 svcb_priority
+        putDomain Canonical svcb_target
```

Comment:
note:

The reason for applying `Canonical` is because

https://datatracker.ietf.org/doc/html/draft-ietf-dnsop-svcb-https-11#section-2.2
"the uncompressed, fully-qualified TargetName"